src/odemis/driver/actuator.py

```diff
@@ -2943,11 +2963,59 @@ def __init__(self, name, role, dependencies, channels, distance, **kwargs):
         # Subscribe to sensor position updates
         self.sensor.position.subscribe(self._on_sensor_position)
 
-    def reference(self):
+        # Executor for referencing
+        self._executor = CancellableThreadPoolExecutor(max_workers=1)
+
+        if ref_on_init == "always":
+            f = self.reference(None, omit_referenced=False)
+        elif ref_on_init in (True, "if necessary"):
+            f = self.reference(None, omit_referenced=True)
+        elif ref_on_init in (False, "never"):
+            f = None
+        else:
+            raise ValueError("Invalid parameter %s for ref_on_init." % ref_on_init)
+
+        if f:
+            f.add_done_callback(self._on_referenced)
+
+    @roattribute
+    def axes(self):
+        """ dict str->Axis: name of each axis available -> their definition."""
+        return self._axes
+
+    @isasync
+    def reference(self, axes, omit_referenced=False):
         """
-        Calls .reference function of sensor.
+        Calls .reference function of stage (if available) and sensor.
+        axes (set of str): axes to be referenced
+        omit_referenced (bool): only reference axes if they are not referenced yet
+        returns (Future): object to control the reference request
         """
-        self.sensor.reference()
+        if not axes:
+            axes = self.channels.keys()
+        f = self._executor.submit(self._doReference, axes, omit_referenced)
+        return f
```

Comment:
You can return right away:
`return self._executor.submit(self._doReference, axes, omit_referenced)`